### PR TITLE
844 unsaved by default

### PR DESCRIPTION
**Pull request by chao-mu**

### CODE REVIEW

File: frontend/src/hooks/useSaveGame.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+} from '@jackstenglein/chess-dojo-common/src/database/game';
+import { useSessionStorage } from 'usehooks-ts';
+
+const STAGED_CREATE_GAME_KEY = 'useSaveGame:stageCreateGame';
+
+export interface UseSaveGameFields {
+    createGame: (req: CreateGameRequest) => Promise<void>;
+    updateGame: (req: UpdateGameRequest) => Promise<void>;
+    stageCreateGame: (req: CreateGameRequest) => void;
+    stagedCreateGame: CreateGameRequest | null;
```
[the remark this answers is not in the document]
Comment (jackstenglein) [reply frontend/src/hooks/useSaveGame.ts:21]: Or change the name of `stageCreateGame`. Maybe something like `stagedRequest` and `setStagedRequest` to match the usual `useState` format?

File: frontend/src/hooks/useSaveGame.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (err) {
+            console.error('CreateGame ', err);
+            request.onFailure(err);
+        }
+
+        if (!data) {
+            return;
+        }
+
+        onCreateGame(createReq, data);
```
Comment (jackstenglein) [inline frontend/src/hooks/useSaveGame.ts:47]: Why is this section outside the try block?

File: frontend/src/hooks/useSaveGame.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        if (!isGame(data)) {
+            request.onSuccess(`Created ${data.count} games`);
+        }
+
+        request.onSuccess();
```
Comment (jackstenglein) [inline frontend/src/hooks/useSaveGame.ts:53]: I think this might need to be in an `else`

File: frontend/src/hooks/useSaveGame.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (err) {
+            console.error('updateGame ', err);
+            request.onFailure(err);
+        }
+
+        if (!data) {
+            return;
+        }
+
+        request.onSuccess();
```
Comment (jackstenglein) [inline frontend/src/hooks/useSaveGame.ts:75]: Why is this outside the try block?

File: frontend/src/hooks/useSaveGame.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        trackEvent(EventType.SubmitGame, {
+            count: 1,
+            method: req.type,
+        });
+
+        const urlSafeId = game.id.replaceAll('?', '%3F');
+        let newUrl = `/games/${game.cohort.replaceAll('+', '%2B')}/${urlSafeId}`;
+
+        if (req.directory) {
+            newUrl += `&directory=${req.directory.id}&directoryOwner=${req.directory.owner}`;
```
Comment (jackstenglein) [inline frontend/src/hooks/useSaveGame.ts:99]: This needs to be `?directory`

File: frontend/src/hooks/useSaveGame.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    const request = useRequest<string>();
+    const [stagedCreateGame, stageCreateGame] =
+        useSessionStorage<CreateGameRequest | null>(STAGED_CREATE_GAME_KEY, null);
+    const { game } = useGame();
+
+    const createGame = async (createReq: CreateGameRequest) => {
+        request.onStart();
+        let data: Game | EditGameResponse | undefined;
+        try {
+            const response = await api.createGame(createReq);
```
Comment (jackstenglein) [inline frontend/src/hooks/useSaveGame.ts:36]: Should we clear the session storage after creating the game? If I create the game and then click back, it reloads the same game in the analysis board and gives me the option to save again, which is kind of strange

File: frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx [hunk 7]
```diff
@@ -32,6 +32,8 @@ import { useNavigate } from 'react-router-dom';
 import { useChess } from '../../../PgnBoard';
 import AnnotationWarnings from '../../../annotations/AnnotationWarnings';
 import RequestReviewDialog from './RequestReviewDialog';
+import { UnsavedGameBanner } from '@/components/games/edit/UnsavedGameBanner';
```
Comment (jackstenglein) [inline frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx:35]: Are these changes necessary? There are no other changes in the file.

**Review by jackstenglein — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/useSaveGame.ts (hunk 4 above)
Comment (chao-mu) [reply frontend/src/hooks/useSaveGame.ts:75]: I avoid unnecessary calls in try { ... }, though I admit here it's a bit confusing.

File: frontend/src/hooks/useSaveGame.ts (hunk 5 above)
Comment (chao-mu) [reply frontend/src/hooks/useSaveGame.ts:99]: Fascinating. I must have somehow changed the character by mistaken.

File: frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx (hunk 7 above)
Comment (chao-mu) [reply frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx:35]: There should be. Something went awry here, the game settings should show a non-dismissible unsaved banner.

Comment (jackstenglein) [reply frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx:35]: It looks like you put it in EditorSettings instead, which probably isn't the correct place for it

**Comment by chao-mu** [comment]
That is what loads for an unsaved game. GameSettings has a bunch of stuff specific for a saved game. I will look into moving it around though.








---- On Sun, 24 Nov 2024 12:15:33 -0500 ***@***.*** wrote ----



@jackstenglein commented on this pull request.

In frontend/src/board/pgn/boardTools/underboard/settings/GameSettings.tsx:

> @@ -32,6 +32,8 @@ import { useNavigate } from 'react-router-dom';
 import { useChess } from '../../../PgnBoard';
 import AnnotationWarnings from '../../../annotations/AnnotationWarnings';
 import RequestReviewDialog from './RequestReviewDialog';
+import { UnsavedGameBanner } from '@/components/games/edit/UnsavedGameBanner';


It looks like you put it in EditorSettings instead, which probably isn't the correct place for it

File: frontend/src/app/(scoreboard)/games/new/NewGameBoard.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (req.pgnText) {
+        return { pgn: req.pgnText };
+    }
+
+    if (req.type === 'startingPosition') {
+        return { fen: FEN_STARTING };
+    }
+
+    return { pgn: '1. a3 h6' };
```
[the remark this answers is not in the document]
Comment (jackstenglein) [reply frontend/src/app/(scoreboard)/games/new/NewGameBoard.tsx]: Still wondering why this is here

**Comment by chao-mu** [comment]
Something didn't make it in then








---- On Tue, 03 Dec 2024 19:32:39 -0500 ***@***.*** wrote ----



@jackstenglein commented on this pull request.

In frontend/src/app/(scoreboard)/games/new/NewGameBoard.tsx:

> +const FEN_STARTING = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1';
+
+function parseCreateGameRequest(req: CreateGameRequest | null) {
+    if (!req) {
+        return { fen: FEN_STARTING };
+    }
+
+    if (req.pgnText) {
+        return { pgn: req.pgnText };
+    }
+
+    if (req.type === 'startingPosition') {
+        return { fen: FEN_STARTING };
+    }
+
+    return { pgn: '1. a3 h6' };


Still wondering why this is here

File: frontend/src/hooks/useSaveGame.ts (hunk 2 above)
Comment (jackstenglein) [reply frontend/src/hooks/useSaveGame.ts:49]: I'd still prefer to have this in the try block. It makes the control flow needlessly complicated to have it outside

File: frontend/src/hooks/useSaveGame.ts (hunk 4 above)
Comment (jackstenglein) [reply frontend/src/hooks/useSaveGame.ts:77]: Same thing with this. `request.onSuccess()` should be inside the try, and we can get rid of the `data` variable

**Review by jackstenglein — APPROVED** [review]
